File: erlang/src/copilot_tools.erl
%%%-------------------------------------------------------------------
%%% @doc Tool definition helpers for the Copilot SDK.
%%%
%%% Provides functions to define tools that the Copilot agent can
%%% invoke. Handles argument parsing, result normalization, and
%%% error wrapping.
%%% @end
%%%-------------------------------------------------------------------
-module(copilot_tools).

-export([
    define_tool/3,
    define_tool/4,
    normalize_result/1
]).

%% @doc Define a tool with a name, description, and handler function.
%% The handler receives a map of parsed arguments and a tool invocation map.
-spec define_tool(binary(), binary(), fun()) -> map().
define_tool(Name, Description, Handler) ->
    define_tool(Name, Description, undefined, Handler).

%% @doc Define a tool with a name, description, JSON Schema parameters, and handler.
-spec define_tool(binary(), binary(), map() | undefined, fun()) -> map().
define_tool(Name, Description, Parameters, Handler) ->
    WrappedHandler = fun(Invocation) ->
        Args = case maps:get(<<"arguments">>, Invocation, #{}) of
            ArgBin when is_binary(ArgBin) ->
                try jsx:decode(ArgBin, [return_maps])
                catch _:_ -> #{}
                end;
            ArgMap when is_map(ArgMap) ->
                ArgMap;
            _ ->
                #{}
        end,
        try Handler(Args, Invocation) of
            Result ->
                normalize_result(Result)
        catch
            _Class:Reason ->
                copilot_types:tool_result(
                    <<"Invoking this tool produced an error. "
                      "Detailed information is not available.">>,
                    failure,
                    iolist_to_binary(io_lib:format("~p", [Reason]))
                )
        end
    end,
    Tool = #{
        <<"name">>        => Name,
        <<"description">> => Description,
        <<"handler">>     => WrappedHandler
    },
    case Parameters of
        undefined -> Tool;
        _         -> Tool#{<<"parameters">> => Parameters}
    end.

%% @doc Normalize a handler return value into a tool result map.
%% Accepts binaries, maps with resultType, other maps (JSON-encoded),
%% and atoms.
-spec normalize_result(term()) -> map().
normalize_result(undefined) ->
    copilot_types:tool_result(<<>>, success);
normalize_result(Result) when is_binary(Result) ->
    copilot_types:tool_result(Result, success);
normalize_result(#{<<"resultType">> := _} = Result) ->
    Result;
normalize_result(Result) when is_map(Result) ->
    copilot_types:tool_result(jsx:encode(Result), success);
normalize_result(Result) when is_list(Result) ->
    copilot_types:tool_result(jsx:encode(Result), success);
normalize_result(Result) ->
    copilot_types:tool_result(
        iolist_to_binary(io_lib:format("~p", [Result])),
        success
    ).

File: fortran/src/copilot_tools.f90
! -------------------------------------------------------------------------------------
!  Copyright (c) Microsoft Corporation. All rights reserved.
! -------------------------------------------------------------------------------------

!> Tool definition helpers for the Copilot SDK.
!! Provides a convenience subroutine for creating tool descriptors.
module copilot_tools
  use copilot_types
  implicit none
  private

  public :: define_tool, tool_success, tool_failure

contains

  !> Create a copilot_tool descriptor.
  !!
  !! @param name          Tool name visible to the assistant
  !! @param description   Human-readable description of what the tool does
  !! @param schema_json   JSON Schema string describing the tool parameters
  !! @param callback      Procedure pointer invoked when the assistant calls the tool
  !! @return              Populated copilot_tool instance
  function define_tool(name, description, schema_json, callback) result(tool)
    character(len=*), intent(in) :: name
    character(len=*), intent(in) :: description
    character(len=*), intent(in) :: schema_json
    procedure(tool_callback_interface) :: callback
    type(copilot_tool) :: tool

    tool%name = name
    tool%description = description
    tool%input_schema_json = schema_json
    tool%callback => callback
  end function define_tool

  !> Create a successful tool result.
  function tool_success(content) result(res)
    character(len=*), intent(in) :: content
    type(tool_result) :: res

    res%content = content
    res%is_error = .false.
  end function tool_success

  !> Create a failure tool result.
  function tool_failure(error_text) result(res)
    character(len=*), intent(in) :: error_text
    type(tool_result) :: res

    res%error_text = error_text
    res%is_error = .true.
  end function tool_failure

end module copilot_tools
